validation/src/signature/pkh_ethereum.rs

```diff
@@ -0,0 +1,259 @@
+use std::str::FromStr;
+
+use anyhow::{bail, Result};
+use k256::ecdsa::RecoveryId;
+use k256::ecdsa::{Signature, VerifyingKey};
+use sha3::Digest;
+use ssi::caip10::BlockchainAccountId;
+use ssi::keccak;
+
+use crate::{cacao::Capability, siwx_message::SiwxMessage};
+
+const ETH_CHAIN: &str = "Ethereum";
+
+#[derive(Debug)]
+pub struct PkhEthereum {}
+
+impl PkhEthereum {
+    /// Verify a cacao generated using SIWE (did:pkh:eip155 with eip4361 capability)
+    pub fn verify(cacao: &Capability) -> anyhow::Result<()> {
+        Self::verify_eip191_signature(cacao)
+    }
+
+    fn verify_eip191_signature(cacao: &Capability) -> anyhow::Result<()> {
+        let issuer = BlockchainAccountId::from_str(&cacao.payload.issuer.replace("did:pkh:", ""))?
+            .account_address
+            .to_lowercase();
+        let siwx = SiwxMessage::from_cacao(cacao)?;
+        // assume the message doesn't use eip55 for the ethereum address
+        let mut recovered =
+            Self::verify_message(&siwx.as_message(ETH_CHAIN), &cacao.signature.signature)
+                .unwrap_or_default();
+
+        if recovered != issuer {
+            // try using eip55 address
+            if let Ok(addr) = keccak::eip55_checksum_addr(&issuer) {
+                recovered = Self::verify_message(
+                    &siwx.as_message_with_address(ETH_CHAIN, &addr),
+                    &cacao.signature.signature,
+                )
+                .unwrap_or_default();
+            }
+        }
+
+        // 2022-09-20 in milliseconds
+        // TODO: use lazy static/const for LEGACY_CHAIN_ID_REORG_DATE
+        if recovered != issuer
+            && cacao.payload.issued_at
+                < chrono::DateTime::from_timestamp_millis(1663632000000).unwrap()
+        {
+            // might be an old CACAOv1 format
+            recovered = Self::verify_message(
+                &siwx.as_legacy_chain_id_message(ETH_CHAIN),
+                &cacao.signature.signature,
+            )
+            .unwrap_or_default();
+        }
+        if recovered != issuer {
+            anyhow::bail!("Signature does not belong to the issuer");
+        }
+        Ok(())
+    }
+
+    fn verify_message(msg: &str, signature: &str) -> Result<String> {
+        // Create the prefixed message as per the Ethereum standard
+        let digest = keccak::hash_personal_message(msg);
+        let signature_bytes = hex::decode(signature.strip_prefix("0x").unwrap_or(signature))?;
+
+        // eth offsets the 0-1 recovery ID by 27 and doesn't use 2 or 3
+        let mut v = signature_bytes[64];
```

Comment:
Should we worry that there aren't 64 bytes?

Reply:
Yeah, I'll add a check. I thought this was being hashed and would end up the right length but that's the other values in this function.